// react-parm/lib/utils.js
'use strict';

exports.__esModule = true;
exports.identity = exports.getNamespacedRef = exports.createRefCreator = exports.logInvalidInstanceError = exports.isClassComponent = exports.bindSetState = exports.addPropTypeIsRequired = exports.IGNORED_STATIC_KEYS = exports.IGNORED_COMPONENT_KEYS = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @constant {Object} IGNORED_COMPONENT_KEYS keys to ignore when creating a component
 */
// external dependencies
var IGNORED_COMPONENT_KEYS = exports.IGNORED_COMPONENT_KEYS = {
  getInitialState: true,
  getInitialValues: true,
  isPure: true,
  onConstruct: true,
  state: true
};

/**
 * @constant {Object} IGNORED_STATIC_KEYS keys to ignore when assigning statics to a component
 */
var IGNORED_STATIC_KEYS = exports.IGNORED_STATIC_KEYS = {
  displayName: true
};

/**
 * @function addPropTypeIsRequired
 *
 * @description
 * add the isRequired method to the propType
 *
 * @param {function} propType the propType checker
 * @returns {function} the propType with the isRequired function added
 */
var addPropTypeIsRequired = exports.addPropTypeIsRequired = function addPropTypeIsRequired(propType) {
  return (propType.isRequired = function (props, key, component) {
    return props[key] == null // eslint-disable-line eqeqeq
    ? new Error('The prop `' + key + '` is marked as required in `' + component + '`, but its value is `' + props[key] + '`.') : propType(props, key, component);
  }) && propType;
};

/**
 * @function bindSetState
 *
 * @description
 * bind the setState method to the component instance to ensure it can be used in a functional way
 *
 * @param {ReactComponent} instance the instance to bind setState to
 * @returns {void}
 */
var bindSetState = exports.bindSetState = function bindSetState(instance) {
  return Object.prototype.hasOwnProperty.call(instance.setState, 'prototype') ? instance.setState = instance.setState.bind(instance) : instance.setState;
};

/**
 * @function isClassComponent
 *
 * @description
 * is the value passed a valid react component class instance
 *
 * @param {any} value the value to test
 * @returns {boolean} is the value a react component instance
 */
var isClassComponent = exports.isClassComponent = function isClassComponent(value) {
  return !!value && value instanceof _react2.default.Component;
};

/**
 * @function logInvalidInstanceError
 *
 * @description
 * notify the user that the instance passed is invalid
 *
 * @param {string} type the type of creator being called
 * @returns {void}
 */
var logInvalidInstanceError = exports.logInvalidInstanceError = function logInvalidInstanceError(type) {
  return console.error('The instance provided for use with the ' + type + ' is not a valid React component instance.');
}; // eslint-disable-line no-console

/**
 * @function createRefCreator
 *
 * @description
 * create a method that will assign a ref value to the instance passed
 *
 * @param {function} getter the function that gets the component value for the ref
 * @returns {function(ReactComponent, string): function((HTMLElement|Component)): void} the ref create
 */
var createRefCreator = exports.createRefCreator = function createRefCreator(getter) {
  return function (instance, ref) {
    return isClassComponent(instance) ? function (component) {
      return instance[ref] = getter(component);
    } : logInvalidInstanceError('ref');
  };
};

/**
 * @function getNamespacedRef
 *
 * @description
 * get the ref that is a combination of the raw component and the component's underlying HTML element
 *
 * @param {ReactComponent} component the component to assin
 * @returns {{component: ReactComponent, element: HTMLElement}} the namespaced ref
 */
var getNamespacedRef = exports.getNamespacedRef = function getNamespacedRef(component) {
  return { component: component, element: (0, _reactDom.findDOMNode)(component) };
};

/**
 * @function identity
 *
 * @description
 * return the first parameter passed
 *
 * @param {any} value the value to pass through
 * @returns {any} the first parameter passed
 */
var identity = exports.identity = function identity(value) {
  return value;
};
