// fast-equals/lib/index.js
'use strict';

exports.__esModule = true;
exports.shallowEqual = exports.deepEqual = exports.circularShallowEqual = exports.circularDeepEqual = exports.sameValueZeroEqual = exports.createCustomEqual = undefined;

var _comparator = require('./comparator');

var _comparator2 = _interopRequireDefault(_comparator);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// comparator
exports.createCustomEqual = _comparator2.default;
exports.sameValueZeroEqual = _utils.sameValueZeroEqual;

// utils

var circularDeepEqual = exports.circularDeepEqual = (0, _comparator2.default)((0, _utils.createCircularEqual)());
var circularShallowEqual = exports.circularShallowEqual = (0, _comparator2.default)((0, _utils.createCircularEqual)(_utils.sameValueZeroEqual));
var deepEqual = exports.deepEqual = (0, _comparator2.default)();
var shallowEqual = exports.shallowEqual = (0, _comparator2.default)(function () {
  return _utils.sameValueZeroEqual;
});

exports.default = {
  circularDeep: circularDeepEqual,
  circularShallow: circularShallowEqual,
  createCustom: _comparator2.default,
  deep: deepEqual,
  sameValueZero: _utils.sameValueZeroEqual,
  shallow: shallowEqual
};
